### PR TITLE
Review annotations and test for allocations in generic matmatmul

### PR DESCRIPTION
**Pull request by dkarrasch**
This adds a few comments as to why compiler annotations are helpful, removes seemingly unhelpful annotations and adds tests to make sure we don't allocate. This is open to further suggestions for tests and/or the return of removed annotations for good reasons.

### CODE REVIEW

**Comment by dkarrasch** [comment]
@jishnub You may wish to review the annotations in the matvecmul case accordingly.

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 1]
```diff
@@ -567,7 +566,7 @@ function gemm_wrapper(tA::AbstractChar, tB::AbstractChar,
     end
 end
 
-Base.@constprop :aggressive function gemm_wrapper!(C::StridedVecOrMat{T}, tA::AbstractChar, tB::AbstractChar,
+function gemm_wrapper!(C::StridedVecOrMat{T}, tA::AbstractChar, tB::AbstractChar,
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/matmul.jl:569]: The other function calls are behind dynamic branches, so it doesn't seem like constants are propagated through to `matmul2x2!`, `matmul3x3!` and `_generic_matmatmul!`, unless I'm mistaken. At least, it seems that calling `gemm_wrapper!` with 2x2 arguments, the whole `matmul2x2!` gets compiled, not only one branch.

**Comment by dkarrasch** [comment]
@nanosoldier `runbenchmarks("linalg", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/52298#issuecomment-1842616018) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f32a9bf_vs_b5abac4/report.md).

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 2]
```diff
@@ -762,8 +761,7 @@ function generic_matmatmul(tA, tB, A::AbstractVecOrMat{T}, B::AbstractMatrix{S})
     generic_matmatmul!(C, tA, tB, A, B)
 end
 
-const tilebufsize = 10800  # Approximately 32k/3
```
Comment (fredrikekre) [inline stdlib/LinearAlgebra/src/matmul.jl:765]: Apparenly ArrayLayouts.jl uses this... Can we keep it for legacy reasons (while also update ArrayLayouts to not use it?)

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/matmul.jl:765]: I added and merged it right away.

**Pull request merged**